src_lockfree/mpsc_queue.ml

```diff
@@ -1,152 +1,145 @@
-    else
-      (* Someone else added a different node first ([p.next] is not [none]).
-         Make [t.tail] more up-to-date, if it hasn't already changed, and try again. *)
-      Node.fold (Atomic.get p.next)
-        ~none:(fun () -> assert false)
-        ~some:(fun p_next ->
-          ignore (Atomic.compare_and_set t.tail p p_next : bool);
-          aux ())
-  in
-  aux ()
-
-let rec push_head t x =
-  let p = t.head in
-  let next = Atomic.get p.next in
-  if next == Node.closed then raise Closed;
-  let node = Node.make ~next x in
-  if Atomic.compare_and_set p.next next (Node.some node) then
-    if
-      (* We don't want to let [tail] get too far behind, so if the queue was empty, move it to the new node. *)
-      next == Node.none
-    then ignore (Atomic.compare_and_set t.tail p node : bool)
-    else
-      ( (* If the queue wasn't empty, there's nothing to do.
-           Either tail isn't at head or there is some [push] thread working to update it.
-           Either [push] will update it directly to the new tail, or will update it to [node]
-           and then retry. Either way, it ends up at the real tail. *) )
-  else (
-    (* Someone else changed it first. This can only happen if the queue was empty. *)
-    assert (next == Node.none);
-    push_head t x)
-
-let rec close (t : 'a t) =
-  (* Mark the tail node as final. *)
-  let p = Atomic.get t.tail in
-  if not (Atomic.compare_and_set p.next Node.none Node.closed) then
-    (* CAS failed because [p] is no longer the tail (or is already closed). *)
-    Node.fold (Atomic.get p.next)
-      ~none:(fun () -> assert false)
-        (* Can't switch from another state to [none] *)
-      ~some:(fun p_next ->
-        (* Make [tail] more up-to-date if it hasn't changed already *)
-        ignore (Atomic.compare_and_set t.tail p p_next : bool);
-        (* Retry *)
-        close t)
+and ('a, _) tail =
+  | Snoc : 'a tail_pack * 'a -> ('a, [> `Snoc ]) tail
+  | Topen : ('a, [> `Topen ]) tail
+  | Tclosed : ('a, [> `Tclosed ]) tail
 
-let pop_opt t =
-  let p = t.head in
-  (* [p] is the previously-popped item. *)
-  let node = Atomic.get p.next in
-  Node.fold node
-    ~none:(fun () -> None)
-    ~some:(fun node ->
-      t.head <- node;
-      let v = node.value in
-      node.value <- Obj.magic ();
-      (* So it can be GC'd *)
-      Some v)
+and 'a tail_pack = T : ('a, [< `Snoc | `Topen | `Tclosed ]) tail -> 'a tail_pack
+[@@unboxed]
 
-exception Empty
+let create () =
+  let tail = Multicore_magic.copy_as_padded @@ Atomic.make (T Topen) in
+  let head = Multicore_magic.copy_as_padded @@ H Hopen in
+  Multicore_magic.copy_as_padded { tail; head }
+
+(* let push t x = t.tail <- T (Snoc (t.tail, x)) *)
+
+let rec push t x =
+  match Atomic.get t.tail with
+  | T Tclosed -> raise Closed
+  | before ->
+      let after = T (Snoc (before, x)) in
+      if not (Atomic.compare_and_set t.tail before after) then push t x
```

Comment:
It would be good to have backoff at this point in case the `compare_and_set` fails (i.e. returns false).

Note that Saturn recently switched to using the [backoff](https://github.com/ocaml-multicore/backoff) library.  If you rebase the branch against `main` you will then get access to the `Backoff` module from the backoff library and can implement backoff to deal with contention on push.